### ISSUE TITLE
eventAllow, eventConstraint, restrict events to resources

### ISSUE THREAD

**Issue by sulian** [fullcalendar/fullcalendar-scheduler#50]
It is possible to constraint an event to a specific type of resource / categorie during an drag'n drop ?

**Comment by alheka** [comment]
I do this by adding an attribute in the events and then check that attribute in eventDrop. Call revertFunc to cancel the drop.

    eventDrop: function(event, delta, revertFunc) {
      if (event.myOwnAttribute === 'canBeDroppedAt') {
        // ...
      } else {
        revertFunc();
      }
    }

**Comment by arshaw** [comment]
i agree, this is a good idea. should somehow be done with [eventContraint](http://fullcalendar.io/docs/event_ui/eventConstraint/) and the like

**Comment by edclements** [comment]
I'm in need of this as well. Here's how I got dropping events to work by modifying fullcalendar but I'd like to know a better way to approach this if there is one.

```Diff
diff --git a/src/EventManager.js b/src/EventManager.js
index 6b007f1..a4b5e69 100644
--- a/src/EventManager.js
+++ b/src/EventManager.js
@@ -1059,7 +1059,12 @@ function EventManager(options) { // assumed to be a calendar
                var eventStart = event.start.clone().stripZone();
                var eventEnd = t.getEventEnd(event).stripZone();
 
-               return range.start >= eventStart && range.end <= eventEnd;
+               if (range.resourceId && event.resourceId) {
+                       return range.start >= eventStart && range.end <= eventEnd &&
+                               range.resourceId == event.resourceId;
+               } else {
+                       return range.start >= eventStart && range.end <= eventEnd;
+               }
        }
 
 
diff --git a/src/common/Grid.events.js b/src/common/Grid.events.js
index b07a067..5697c2d 100644
--- a/src/common/Grid.events.js
+++ b/src/common/Grid.events.js
@@ -390,6 +390,9 @@ Grid.mixin({
                                end: null, // end should be cleared
                                allDay: !dragEnd.hasTime()
                        };
+                       if (startCell.resourceId) {
+                               dropLocation.resourceId = startCell.resourceId;
+                       }
                }
 
                return dropLocation;
```

**Comment by wespen** [comment]
+1 to add it to eventContraint

**Comment by arshaw** [comment]
In the core project, I am proposing there be a callback called `eventAllow`
https://github.com/fullcalendar/fullcalendar/issues/2740

One of the argument will be a `resource`.

I'm proposing the API for this issue be possible via `eventAllow` and doing conditional behavior based on this resource, either returning `true` or `false` as to whether the drop is allowed.

**Comment by arshaw** [comment]
Example:

```js
eventAllow: function(event, start, end, resource) {
    return true; // or false
}
```

**Comment by arshaw** [comment]
It would also be convenient to allow [eventConstraint](http://fullcalendar.io/docs/event_ui/eventConstraint/) to specify a `resourceId`. Example:

```js
eventConstraint: {
    resourceId: 'a'
}
```

Not super useful as a global option, but very useful as an event object options.

**Comment by arshaw** [comment]
should short-circuit and never execute `eventAllow` if changing resources and [eventResourceEditable](https://github.com/fullcalendar/fullcalendar-scheduler/issues/140) is `false`

**Comment by arshaw** [comment]
note to self: `eventContraint:'businessHours'` will need to be compatible with [business hours per-resource](https://github.com/fullcalendar/fullcalendar-scheduler/issues/61) if it is implemented

**Comment by arshaw** [comment]
Should `eventConstraint` accept multiple resource ID's? `resourceIds: [ ]` i think yes

**Comment by mitomm** [comment]
👍 for the feature and doing it with eventConstraint. Using `resourceIds: [ ]` similar to `dow: [ ]` would be nice.

**Comment by NiklasKamenisch** [comment]
Is there a rough schedule for these adjustments? I could use it for one or the other project.

**Comment by KrunchMuffin** [comment]
+1 would be ideal for me as well.

**Comment by pieterroos** [comment]
Would be great if this would be possible

**Comment by arshaw** [comment]
This has been implemented in v1.4.0!

**Comment by arshaw** [comment]
blog post: https://fullcalendar.io/blog/2016/09/feature-packed-releases/

`eventConstraint` now accepts resource IDs:
https://fullcalendar.io/docs/resource_events/eventConstraint/

new `eventAllow` callback for programmatic control:
https://fullcalendar.io/docs/resource_events/eventAllow/
